### PR TITLE
feat(frontend): support grouping sets

### PR DESCRIPTION
**Pull request by chenzl25**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

- Related Issue https://github.com/risingwavelabs/risingwave/issues/9284
- Currently, the group by clause can only support one and only one grouping sets expression.
- The optimizer use `GroupingSetsToExpandRule` to convert an Agg with grouping sets into an Agg without grouping sets.
- The schema of an Agg with grouping sets is the union of all grouping columns pulsing the agg calls.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR contains user-facing changes.

<details><summary>Click here for Documentation</summary>

### Types of user-facing changes

Please keep the types that apply to your changes, and remove the others.

- Installation and deployment
- Connector (sources & sinks)
- SQL commands, functions, and operators
- RisingWave cluster configuration changes
- Other (please specify in the release note below)

### Release note

</details>

### CODE REVIEW

**Comment by TennyZhuang** [comment]
Can you add Sqlsmith generation support?

**Comment by chenzl25** [comment]
> Can you add Sqlsmith generation support?

Sure, I will reach out to @kwannoel for more information about sqlsmith tests later.

File: e2e_test/batch/aggregate/grouping_sets.slt.part [hunk 1]
```diff
@@ -0,0 +1,20 @@
+statement ok
+SET RW_IMPLICIT_FLUSH TO true;
+
+statement ok
+create table items_sold (brand varchar, size varchar, sales int);
+
+statement ok
+insert into items_sold values ('Foo', 'L', 10),('Foo', 'M', 20),('Bar', 'M', 15),('Bar', 'L', '5');
+
+query IIII rowsort
```
Comment (BugenZhao) [inline e2e_test/batch/aggregate/grouping_sets.slt.part:10]:
```suggestion
query TTII rowsort
```

File: src/frontend/planner_test/tests/testdata/input/grouping_sets.yaml [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+  sql: |
+    create table items_sold (brand varchar, size varchar, sales int);
+    SELECT brand, size, sum(sales) FROM items_sold GROUP BY GROUPING SETS ((brand), (size), ());
+  expected_outputs:
+    - batch_plan
+    - stream_plan
+- name: with distinct
+  sql: |
+    create table items_sold (brand varchar, size varchar, sales int);
+    SELECT brand, size, sum(sales) FROM items_sold GROUP BY GROUPING SETS ((brand), (size), ());
```
Comment (BugenZhao) [inline src/frontend/planner_test/tests/testdata/input/grouping_sets.yaml:11]: Seems no difference for this query?

File: src/frontend/src/planner/select.rs [hunk 3]
```diff
@@ -98,9 +99,9 @@ impl Planner {
         // Plan the SELECT clause.
         // TODO: select-agg, group-by, having can also contain subquery exprs.
         let has_agg_call = select_items.iter().any(|expr| expr.has_agg_call());
-        if !group_by.is_empty() || having.is_some() || has_agg_call {
+        if !group_by.is_empty() || !grouping_sets.is_empty() || having.is_some() || has_agg_call {
             (root, select_items, having) =
-                LogicalAgg::create(select_items, group_by, having, root)?;
+                LogicalAgg::create(select_items, group_by, grouping_sets, having, root)?;
```
Comment (BugenZhao) [inline src/frontend/src/planner/select.rs:104]: Is either `group_by` non-empty or `grouping_set` non-empty? If so, what about making it an enum in `LogicalAggBuilder` and `BoundSelect`?

File: src/frontend/planner_test/tests/testdata/input/grouping_sets.yaml (hunk 2 above)
Comment (chenzl25) [reply src/frontend/planner_test/tests/testdata/input/grouping_sets.yaml:11]: Oops, miss the distinct.

File: src/frontend/src/planner/select.rs (hunk 3 above)
Comment (chenzl25) [reply src/frontend/src/planner/select.rs:104]: Yes, that is true for `LogicalAggBuilder` and `BoundSelect`.

File: src/frontend/src/utils/group_by.rs [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+
+impl GroupBy {
+    pub fn is_empty(&self) -> bool {
+        match self {
+            GroupBy::GroupKey(group_key) => group_key.is_empty(),
+            GroupBy::GroupingSets(grouping_sets) => grouping_sets.is_empty(),
+        }
+    }
+
+    pub fn iter(&self) -> Box<dyn Iterator<Item = &ExprImpl> + '_> {
```
Comment (BugenZhao) [inline src/frontend/src/utils/group_by.rs:31]: nit: we can use `#[auto_enum(Iterator)]` for this.

**Review by st1page — APPROVED** [review]
generally LGTM

File: src/frontend/src/utils/group_by.rs (hunk 4 above)
Comment (chenzl25) [reply src/frontend/src/utils/group_by.rs:31]: That's cool. Let me refactor by `#[auto_enum(Iterator)]`.

**Pull request merged**